src/storage-client/src/controller.rs

```diff
@@ -2383,11 +2385,15 @@ mod collection_mgmt {
                                     diff,
                                 }).collect::<Vec<_>>(), T::from(now()))];
 
-                                // TODO? Handle contention among multiple writers
-                                write_handle.monotonic_append(updates)
-                                    .await
-                                    .expect("sender hung up")
-                                    .expect("no write contention on collections");
+                                loop {
```

Comment:
It's dangerous to blindly retry here. We don't know if this failed because of an indeterminate error, but we _did_ succeed in writing, so retrying here might lead to duplicates.

cc @danhhz who thought more deeply about the problems of transactions and indeterminate errors. This might become easier once we remove indeterminate errors from the `compare_and_append` API.

Reply:
Mmm, interesting. Duplicates are fine for this particular use case, but I appreciate we may not be willing to do it in general.

Maybe we can either make the behaviour configurable, or if it happens that all current users are okay with the risk of duplicates, just put a scary comment?

Reply:
I've pushed a narrow retry loop here: we now only retry on invalid-uppers, which indicate that someone else appended to them first, and should always be safe to retry in the case of monotonic appends. Does that seem any better?

Reply:
I think we need to chat about this in person! I'm skeptical about the correctness, but I might be wrong.

Reply:
Discussed! Aljoscha was correct to be skeptical.

I had thought that `StorageError::InvalidUppers` would only occur when we received an `uppers` error from persist, but it appears we also return it after an indeterminate error when the uppers have changed. (Which I should have remembered; I was poking at the relevant code recently.) Retrying in that case could introduce a duplicate. Which, in our case I think is fine but is not safe behaviour in general. Will need to think on this one more!

Reply:
> I had thought that StorageError::InvalidUppers would only occur when we received an uppers error from persist, but it appears we also return it after an indeterminate error when the uppers have changed. (Which I should have remembered; I was poking at the relevant code recently.) Retrying in that case could introduce a duplicate. Which, in our case I think is fine but is not safe behaviour in general. Will need to think on this one more!

Can you expand on this? I'm not quite sure I follow

Reply:
Sure, though to be clear this is not at all persist's fault so feel free to ignore.

Relevant code stanza: https://github.com/MaterializeInc/materialize/blob/8004b6ee5d34e63b465917fc44254de854e44e0c/src/storage-client/src/controller.rs#L2169-L2182

Effectively two ways we may return a `Err(gid)` here, which will result in an `InvalidUppers` error below:
- If persist returns an `uppers` error in the innermost result. Good, expected.
- If persist returns an indeterminate error, and we notice that the uppers don't match the expected before or after. Regrettable!
And notably `Ok` can either mean success as usual, or that we failed with an indeterminate error but the upper seems to have advanced as expected. In both cases, the caller doesn't have enough info to make a good decision.

One way to fix this is to push more information into the error type, but also we may just wait for indeterminate errors to go away. We'll see.

Reply:
ah, ty! I didn't notice that storage had its own InvalidUppers error, but now I see that it was blatantly obvious above if I'd read the thread more carefully 🙈 

context: a recent change went into persist that involved subtle logic around an upper mismatch and Indeterminate, except we do the exact opposite (an indeterminate followed by a upper mismatch on retry is turned into an indeterminate for Reasons) so my ears perked up when I thought maybe I'd somehow flipped them.

Reply:
This should be resolved now that indeterminate errors are gone!

Arguably this retry loop should be pushed down to the background thread, since all callers of monotonic-append should be comfortable retrying upper errors. (Since by definition they don't care about the current upper.) More than I want to bite off here, though!